crates/bevy_render/src/batching/gpu_preprocessing.rs

```diff
@@ -166,53 +368,123 @@ pub fn delete_old_work_item_buffers<GFBD>(
 pub fn batch_and_prepare_sorted_render_phase<I, GFBD>(
-    gpu_batched_instance_buffers: ResMut<
-        BatchedInstanceBuffers<GFBD::BufferData, GFBD::BufferInputData>,
-    >,
-    mut views: Query<(Entity, &mut SortedRenderPhase<I>)>,
-    param: StaticSystemParam<GFBD::Param>,
+    gpu_array_buffer: ResMut<BatchedInstanceBuffers<GFBD::BufferData, GFBD::BufferInputData>>,
+    mut indirect_parameters_buffer: ResMut<IndirectParametersBuffer>,
+    mut views: Query<(Entity, &mut SortedRenderPhase<I>, Has<GpuCulling>)>,
+    system_param_item: StaticSystemParam<GFBD::Param>,
 ) where
     I: CachedRenderPipelinePhaseItem + SortedPhaseItem,
     GFBD: GetFullBatchData,
 {
-    let system_param_item = param.into_inner();
-
     // We only process GPU-built batch data in this function.
     let BatchedInstanceBuffers {
         ref mut data_buffer,
         ref mut work_item_buffers,
         ..
-    } = gpu_batched_instance_buffers.into_inner();
+    } = gpu_array_buffer.into_inner();
 
-    for (view, mut phase) in &mut views {
+    for (view, mut phase, gpu_culling) in &mut views {
         // Create the work item buffer if necessary.
-        let work_item_buffer = work_item_buffers
-            .entry(view)
-            .or_insert_with(|| BufferVec::new(BufferUsages::STORAGE));
+        let work_item_buffer =
+            work_item_buffers
+                .entry(view)
+                .or_insert_with(|| PreprocessWorkItemBuffer {
+                    buffer: BufferVec::new(BufferUsages::STORAGE),
+                    gpu_culling,
+                });
 
-        super::batch_and_prepare_sorted_render_phase::<I, GFBD>(&mut phase, |item| {
-            let (input_index, compare_data) =
-                GFBD::get_index_and_compare_data(&system_param_item, item.entity())?;
-            let output_index = data_buffer.add() as u32;
+        // Walk through the list of phase items, building up batches as we go.
+        let mut batch: Option<SortedRenderBatch<GFBD>> = None;
+        for current_index in 0..phase.items.len() {
+            // Get the index of the input data, and comparison metadata, for
+            // this entity.
+            let current_batch_input_index = GFBD::get_index_and_compare_data(
+                &system_param_item,
+                phase.items[current_index].entity(),
+            );
+
+            // Unpack that index and metadata.
+            let (current_input_index, current_meta);
+            match current_batch_input_index {
+                Some((input_index, Some(current_compare_data))) => {
+                    current_input_index = Some(input_index);
+                    current_meta = Some(BatchMeta::new(
+                        &phase.items[current_index],
+                        current_compare_data,
+                    ));
+                }
+                _ => {
+                    current_input_index = None;
+                    current_meta = None;
+                }
```

Comment:
Should `Some((input_index, None))` end up here?

Reply:
Yes, because if `CompareData` is `None`, then the entity is unbatchable, and therefore we should ignore it after breaking the batch. I added a comment to that effect.